precompiles/erc20/integration_test.go

```diff
@@ -0,0 +1,1151 @@
+			const (
+				denom       = "uxmpl"
+				expName     = "Xmpl"
+				expSymbol   = "XMPL"
+				expDecimals = uint8(6)
+			)
+
+			BeforeEach(func() {
+				// Deploy new precompile for this test using the xmpl ibc voucher
+				//
+				// NOTE: this is not the same as the one used in the other tests
+				// because we need to set the metadata for this one.
+				ibcDenomTrace := types.DenomTrace{Path: "channel-0", BaseDenom: denom}
+
+				// overwrite the other precompile with this one, so that the test utils like s.getTxAndCallArgs still work.
+				s.precompile = s.setupERC20Precompile(ibcDenomTrace.IBCDenom())
+
+				// TODO: can I handle this differently, just using the integration utils and not using the keeper directly?
+				s.network.App.TransferKeeper.SetDenomTrace(s.network.GetContext(), ibcDenomTrace)
+			})
+
+			DescribeTable("querying the name should return the name", func(callType int) {
+				txArgs, nameArgs := s.getTxAndCallArgs(directCall, contractAddr, erc20.NameMethod)
+
+				_, ethRes, err := s.factory.CallContractAndCheckLogs(sender.Priv, txArgs, nameArgs, passCheck)
+				Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+				var name string
+				err = s.precompile.UnpackIntoInterface(&name, erc20.NameMethod, ethRes.Ret)
+				Expect(err).ToNot(HaveOccurred(), "failed to unpack result")
+				Expect(name).To(Equal(expName), "expected different name")
+			},
+				Entry(" - direct call", directCall),
+				Entry(" - through contract", contractCall),
+			)
+
+			DescribeTable("querying the symbol should return the symbol", func(callType int) {
+				txArgs, symbolArgs := s.getTxAndCallArgs(directCall, contractAddr, erc20.SymbolMethod)
+
+				_, ethRes, err := s.factory.CallContractAndCheckLogs(sender.Priv, txArgs, symbolArgs, passCheck)
+				Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+				var symbol string
+				err = s.precompile.UnpackIntoInterface(&symbol, erc20.SymbolMethod, ethRes.Ret)
+				Expect(err).ToNot(HaveOccurred(), "failed to unpack result")
+				Expect(symbol).To(Equal(expSymbol), "expected different symbol")
+			},
+				Entry(" - direct call", directCall),
+				Entry(" - through contract", contractCall),
+			)
+
+			DescribeTable("querying the decimals should return the decimals", func(callType int) {
+				txArgs, decimalsArgs := s.getTxAndCallArgs(callType, contractAddr, erc20.DecimalsMethod)
+
+				_, ethRes, err := s.factory.CallContractAndCheckLogs(sender.Priv, txArgs, decimalsArgs, passCheck)
+				Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+				var decimals uint8
+				err = s.precompile.UnpackIntoInterface(&decimals, erc20.DecimalsMethod, ethRes.Ret)
+				Expect(err).ToNot(HaveOccurred(), "failed to unpack result")
+				Expect(decimals).To(Equal(expDecimals), "expected different decimals")
+			},
+				Entry(" - direct call", directCall),
+				// FIXME: this is failing??
+				Entry(" - through contract", contractCall),
```

Comment:
not sure why this is failing currently, since the other metadata queries work fine.. 🤔